[SUREFIRE-1994] Upgrade javacc-maven-plugin, add setup instructions for Eclipse

sman-81 opened the pull request:
This PR addresses the issue in [SUREFIRE-1994] to upgrade/configure `javacc-maven-plugin`.

✅ Each commit in the pull request should have a meaningful subject line and body.
✅ Format the pull request title like `[SUREFIRE-XXX] - Fixes bug in ApproximateQuantiles`
✅ Run `mvn clean install` to make sure basic checks pass.
(I use `mvn clean verify` usually which is the same but does not put artifacts into the local repo.)

✅ I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

slawekjaranowski commented on surefire-grouper/pom.xml:
```diff
@@ -32,18 +32,53 @@
             </goals>
+            <configuration>
+              <sourceDirectory>${project.basedir}/src/main/javacc</sourceDirectory>
+              <outputDirectory>${generated.sources.dir}</outputDirectory>
```
why change default value .... `${project.build.directory}/generated-sources/javacc`

sman-81 replied:
It's a bad default imho. People don't expect source folders to exist in the build directory.

Tibor17 replied:
@sman-81 
Don;t modify default settings unless it is absolutely necessary.
The old plugin is compliant with IDEs, the default layout, where generates sources appear in `target/generated-sources/javacc`.

slawekjaranowski commented on surefire-grouper/pom.xml:
```diff
@@ -32,18 +32,53 @@
               <goal>javacc</goal>
             </goals>
+            <configuration>
+              <sourceDirectory>${project.basedir}/src/main/javacc</sourceDirectory>
```
default value is `${basedir}/src/main/javacc`, so looks the same

sman-81 replied:
I like things to be obvious and explicit. Devs not familiar with the workings of this plugin will be able to understand better.

Tibor17 replied:
@sman-81 
Don;t modify default settings unless it is absolutely necessary.
The old plugin is compliant with IDEs, the default layout, where generates sources appear in `target/generated-sources/javacc`.
The `src/main` files are a subject to git commit them. So I disagree to touch this folder.

Tibor17 replied:
Nowadays the IDEs can be configured to run the compilation and lokkup `target/generated-sources` which is default path.

sman-81 replied:
> The `src/main` files are a subject to git commit them. So I disagree to touch this folder.

@Tibor17 
Did you notice `.gitignore` was updated to prevent this?

Take a look at my comment in the Jira-Ticket.

Tibor17 commented:
I have checked it out and the following plugin fails on error
`Execution javacc of goal org.javacc.plugin:javacc-maven-plugin:3.0.3:javacc failed: A required class was missing while executing org.javacc.plugin:javacc-maven-plugin:3.0.3:javacc: org/javacc/parser/Main`

```
              <groupId>org.javacc.plugin</groupId>
              <artifactId>javacc-maven-plugin</artifactId>
              <version>3.0.3</version>
```

I have used another plugin which release was made 20 days ago and it is stable
```
          <groupId>com.helger.maven</groupId>
          <artifactId>ph-javacc-maven-plugin</artifactId>
          <version>4.1.4</version>
```

The directory structure is the same and I did not provide any additional configuration (only used groupId:artifactId:version).

sman-81 commented:
@Tibor17 for `org.javacc.plugin:javacc-maven-plugin` you need to configure `javacc` as a dependency. It's in my pull request.

Tibor17 commented:
This works as well and it is binary compliant with the old plugin.
```
      <plugin>
          <groupId>org.javacc.plugin</groupId>
          <artifactId>javacc-maven-plugin</artifactId>
          <version>3.0.3</version>
        <executions>
          <execution>
            <id>javacc</id>
            <goals>
              <goal>javacc</goal>
                <goal>jjdoc</goal>
            </goals>
          </execution>
        </executions>
          <dependencies>
              <dependency>
                  <groupId>net.java.dev.javacc</groupId>
                  <artifactId>javacc</artifactId>
                  <version>7.0.10</version>
              </dependency>
          </dependencies>
      </plugin>
```

sman-81 commented:
Hello @Tibor17, @slawekjaranowski,
Updated the pr to use defaults in javacc-maven-plugin, removed build-helper-maven-plugin, add added setup instructions for development in Eclipse.

Tibor17 commented:
@sman-81 
It looks fine so far but I would like to review only one commit. Can you pls squash both commits? Thx
BTW, how is your Eclipse settings now? Did my links help in Jira?

sman-81 commented:
@Tibor17

I've merged upstream modifications, git-squashed the two commits of this ticket, and renamed the PR.

Tibor17 commented on surefire-grouper/pom.xml:
```diff
@@ -32,12 +32,16 @@
   <description>Maven Surefire Test-Grouping Support</description>
 
+  <properties>
+    <javacc.version>7.0.10</javacc.version>
```
You can inline this value, that's not a problem for us, we will understand the point anyway.

sman-81 replied:
OK

Tibor17 commented:
@sman-81 
You used `git rebase` or you have used `git merge`?
It looks like you use merge and that's the reason why we can see commits of master as yours.
It would be maybe easier for you cherry pick only your commit and use it in a fresh PR. I am afraid this has conflicts with master it the GH does not let be merge this PR - the button is shaded.

sman-81 commented:
> @sman-81 You used `git rebase` with `forced push` or you have used `git merge`?

I fetched upstream at github on my branch
then pulled
squashed using `git rebase -i HEAD~15`
followed by `push --force`

Tibor17 commented:
@sman-81 
yes, feel free to go for it. Thx

sman-81 commented:
Hello @Tibor17

now the pr is correctly rebased and squashed. Please take a look :)
Thanks!

Tibor17 commented on README.md:
````diff
@@ -63,7 +63,23 @@ Build the Surefire project using **Maven 3.2.5+** and **JDK 1.8+**.
+  * Maven-build (without tests) the `surefire-shared-utils`  module with profile `ide-development` and install it into the local maven repository
+    ```
+    cd surefire-shared-utils
+    mvn clean install -Pide-development -DskipTests
````
The module does not have tests. More simple command line is `mvn install -P ide-development -f surefire-shared-utils/pom.xml`.

Tibor17 commented on README.md:
````diff
@@ -63,7 +63,23 @@ Build the Surefire project using **Maven 3.2.5+** and **JDK 1.8+**.
+  * Maven-build the `surefire-grouper` module to generate sources into `target/generated-sources/javacc`
+    ```
+    cd surefire-grouper
+    mvn clean verify
````
Perhaps rephrase the sentense to:
Build the module `surefire-grouper` using the Maven command `mvn compile -f surefire-grouper/pom.xml` in order to generate and compile sources into `target/generated-sources/javacc`.

sman-81 replied:
@Tibor17 
> Perhaps rephrase the sentense ...

Done :)

Tibor17 commented:
@sman-81
@slawekjaranowski
LGTM. No objections?

sman-81 commented:
> @sman-81 @slawekjaranowski LGTM. No objections?

well ..., I don't :)
have a good week-end!

Tibor17 commented:
Thx for contributing and the PR! Let's keep tehm coming more.

sman-81 commented:
@Tibor17  Thank you, too. I like your attention to detail. It was good working together!
